pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentDispatcherMultipleConsumers.java

```diff
@@ -0,0 +1,195 @@
+        if (consumerSet.removeAll(consumer) == 1) {
+            consumerList.remove(consumer);
+            log.info("Removed consumer {}", consumer);
+            if (consumerList.isEmpty()) {
+                if (closeFuture != null) {
+                    log.info("[{}] All consumers removed. Subscription is disconnected", name);
+                    closeFuture.complete(null);
+                }
+                TOTAL_AVAILABLE_PERMITS_UPDATER.set(this, 0);
+            }
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Trying to remove a non-connected consumer: {}", name, consumer);
+            }
+            TOTAL_AVAILABLE_PERMITS_UPDATER.addAndGet(this, -consumer.getAvailablePermits());
+        }
+    }
+
+    @Override
+    public boolean isConsumerConnected() {
+        return !consumerList.isEmpty();
+    }
+
+    @Override
+    public CopyOnWriteArrayList<Consumer> getConsumers() {
+        return consumerList;
+    }
+
+    @Override
+    public synchronized boolean canUnsubscribe(Consumer consumer) {
+        return consumerList.size() == 1 && consumerSet.contains(consumer);
+    }
+
+    @Override
+    public CompletableFuture<Void> close() {
+        IS_CLOSED_UPDATER.set(this, TRUE);
+        return disconnectAllConsumers();
+    }
+
+    @Override
+    public synchronized void consumerFlow(Consumer consumer, int additionalNumberOfMessages) {
+        if (!consumerSet.contains(consumer)) {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Ignoring flow control from disconnected consumer {}", name, consumer);
+            }
+            return;
+        }
+
+        totalAvailablePermits += additionalNumberOfMessages;
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Trigger new read after receiving flow control message", consumer);
+        }
+    }
+
+    @Override
+    public synchronized CompletableFuture<Void> disconnectAllConsumers() {
+        closeFuture = new CompletableFuture<>();
+        if (consumerList.isEmpty()) {
+            closeFuture.complete(null);
+        } else {
+            consumerList.forEach(Consumer::disconnect);
+        }
+        return closeFuture;
+    }
+
+    @Override
+    public void reset() {
+        IS_CLOSED_UPDATER.set(this, FALSE);
+    }
+
+    @Override
+    public SubType getType() {
+        return SubType.Shared;
+    }
+
+    @Override
+    public void sendMessages(List<Entry> entries) {
+        Consumer consumer = TOTAL_AVAILABLE_PERMITS_UPDATER.get(this) > 0 ? getNextConsumer() : null;
+        if (consumer != null) {
+            TOTAL_AVAILABLE_PERMITS_UPDATER.addAndGet(this, -consumer.sendMessages(entries).getRight());
+        } else {
+            entries.forEach(Entry::release);
+        }
+    }
+
+    @Override
+    public boolean hasPermits() {
+        return TOTAL_AVAILABLE_PERMITS_UPDATER.get(this) > 0;
+    }
+
+    private Consumer getNextConsumer() {
```

Comment:
Great, you can introduce some base class like `AbstractDispatcherMultipleConsumers` with the common logic.